File: algorithms/sort/cocktailSort.go
package sort

func CocktailSort(arr []int) []int {
	rightBorder := len(arr) - 1
	leftBorder := 0
	rIndex := 0
	lIndex := 0
	for i := 0; i < len(arr)/2-1; i++ {
		Flag := true
		for j := i; j < rightBorder; j++ {
			if arr[j] > arr[j+1] {
				arr[j], arr[j+1] = arr[j+1], arr[j]
				Flag = false
				rIndex = j
			}
		}
		rightBorder = rIndex
		if Flag {
			break
		}

		for j := rightBorder; j > leftBorder; j-- {
			if arr[j] < arr[j-1] {
				arr[j], arr[j-1] = arr[j-1], arr[j]
				Flag = false
				lIndex = j
			}
		}
		leftBorder = lIndex
		if Flag {
			break
		}
	}
	return arr
}

File: sort_dart/cocktailSort.dart
// 鸡尾酒排序是一个双向冒泡的算法.

List cocktailSort(List arr) {
  var len = arr.length;

  bool isSorted = false;
  for (int i = 0; i < len / 2; i++) {
    isSorted = false;

    for (int j = i; j < len - i - 1; j++) {
      if (arr[j] > arr[j + 1]) {
        var tmp = arr[j];
        arr[j] = arr[j + 1];
        arr[j + 1] = tmp;

        isSorted = true;
      }
    }

    for (int j = len - i - 1; j > i; j--) {
      if (arr[j] < arr[j - 1]) {
        var tmp = arr[j];
        arr[j] = arr[j - 1];
        arr[j - 1] = tmp;

        isSorted = true;
      }
      isSorted = true;
    }

    if (isSorted) break;
  }

  return arr;
}

void main(List<String> args) {
  print(cocktailSort([2, 4, 6, 1, 4, 3, 5, 9, 7]));
}
